### ISSUE TITLE
⚠️ Nextcloud 28 support ⚠️

### ISSUE THREAD

**Issue by te-online** [te-online/files_linkeditor#92]
There's a plan now! 🎉 

- [x] Get the app to build and run again with the old and new JS dependencies it requires (especially `svelte` and `@nextcloud/files`) ⌛ 
- [x] Implement [`addNewFileMenuEntry`](https://nextcloud-libraries.github.io/nextcloud-files/functions/addNewFileMenuEntry.html) to replace `window.OC.Plugins.register("OCA.Files.NewFileMenu", ...)`
- [x] Implement [`registerFileAction`](https://nextcloud-libraries.github.io/nextcloud-files/functions/registerFileAction.html) to replace `window.OCA.Files.fileActions.registerAction` for viewing and editing
- [x] ~~Implement public share "view only" functionality in new files list~~ (new file list seems not to be used on public pages, yet)
- [x] Add default action when clicking on the file using the corresponding new API (???)
- [ ] Testing
- [ ] Testing in NC 26 and 27 to determine if the new version will need to cut off support to these NC versions

----

**tl;dr: Don't upgrade to Nextcloud 28 (yet) if you absolutely need this app.**

I've been looking at breaking changes in Nextcloud 28 and there's a lot affecting this app.

The JavaScript API for the files app has essentially been dropped without officially deprecating it (or maybe I've simply missed that step).

Regardless, there is some work to do, calling the new APIs to register file handlers and integrating with the dav client.

It will take some time, please bear with me 😊 Contributions welcome.

**Comment by jacotec** [comment]
@te-online 
Hi Thomas,
any update in regards to NC 28 compatibility?

**Comment by larsik78** [comment]
We count on you!

**Comment by arkascha** [comment]
Maybe it makes sense to contact the nextcloud dev Team directly about this. 
If that JS API really got dropped without prior notice, then I could imagine that people were unaware about the consequences. And that there is motivation to make up for such an issue that actually affects quite a lot of people in an annoying manner. Especially _because_ it came completely unannounced to the user base.

**Comment by te-online** [comment]
The team at Nextcloud is moving on with the Files app and I generally appreciate that. 

However, it would have been more appropriate to release a Beta Files app for 28 and then introduce the new app in 29, so everyone has a chance to get ready. There may have been opportunities to get this ready early, but I did not hear of any prior to the release of 28 Beta.

After all, files are the core functionality of Nextcloud and all integrations must be re-written, because the architecture has changed completely. Additionally, I've been quite pressed for time lately, adding to the delay in regards to this app.

That being said, I don't think the core team has any interest in bringing back to old and dated files APIs and we just have to deal with this now and hope for less "move fast and break things" in the future.

**Comment by jacotec** [comment]
Thank you for the insight @te-online ! This is the first time ever since using Nextcloud that we're at a .3 release with even the following main in Beta and I still can't upgrade because essential apps (like yours) do not work.

Thanks for working to make it compatible!

**Comment by te-online** [comment]
I made some huge progress on this, but am waiting for some help for adding the "new file" entries. There is a lack of documentation and I couldn't figure it out on my own, nor did I see any examples in other apps.

Once I figure out how to add the "new file" entries properly, there will be an update compatible with NC28 🙏 😊 

Just a note: While developing I noticed quite a few glitches in the new files app in general – so maybe hold off on updating to 28 either way (I'm using the stable version of 28)...

**Comment by te-online** [comment]
Functionality is now in place for NC28, I've been preparing this on a feature branch, which I will merge after testing.

**Comment by arkascha** [comment]
Does not work for me. 

I updated the "Link Editor" app, it is enabled. Also disabled and reenabled it in a Nextcloud 28.0.3 installation. 
I do not see an option to create such files, nor an option to edit them. When clicking the the file is downloaded.

**Comment by te-online** [comment]
Thanks for reporting! I will have a look and come back with an update shortly.

**Comment by te-online** [comment]
@arkascha Quick question: Is the file on your server in this location? `files_linkeditor/dist/bundle.js`? It works perfectly fine in my docker setup, so one problem could be that your app installation for some reason doesn't include the file.

**Comment by te-online** [comment]
I reverted to putting the bundle file in the `js` folder where Nextcloud seems to expect it to be, to avoid weird directory traversal, which seems to break in your case. I'm just going to go ahead and release it as 1.1.18 in the hopes that this will resolve it for good 🙏

**Comment by arkascha** [comment]
I tested again, this is progress! 
I do have the options to create ".webloc" or ".url" files now. The edit dialog works, files are created and usable. 

However there are a few issues I observed: 

- The file name extension is not added automatically to the name. Why not, if one explicitly has to chose it in the step before?
- Without specifying the file name extension nextcloud does not recognize the files (like in the 1980th it still appears to rely on "file name extensions", which is not the case any more everywhere else? This is probably not an issue with this app, but with the nextcloud core. 
- Renaming a file named without file name extension does not result in it being recognized apparently, the file icon does not change, not even when reloading the page. 
- Files are created in the wrong folder: one level up! So not inside the folder where one chooses the menu entry to create them. Why?!? One has to move each one afterwards. 
- There is no validation of entered links at all. Only when using the files one gets a hint that the file appears to be broken. 

So the app is usable again. That is great! But it needs some love to smooth things out. Especially the issue that files are created in the wrong folder is a major clear bug.

**Comment by te-online** [comment]
Hey, thanks for testing 🙌 I should really create a beta channel for this app... 

> The file name extension is not added automatically to the name. Why not, if one explicitly has to chose it in the step before?

Yes, this is a limitation of the very basic filename modal I was forced to make. The "text" app, which is included in Nextcloud core, does the same, however, I tried to mimic the behavior for a first version.

Unfortunately, the "Files" app maintainers did not keep their API stable, so for Nextcloud 28 I had to work with the new APIs. [I have criticized this](https://github.com/nextcloud/server/issues/35503#issuecomment-1987317724), mainly because it cost me a lot of time to fix something that already worked with the old API. One of the changes they chose to make was leaving the file naming and creation to each app integrating with the "New file" menu, which means each app has to fix their own UX issues with file naming and creation. This is one of those issues and I will try to address it eventually.

> Without specifying the file name extension nextcloud does not recognize the files (like in the 1980th it still appears to rely on "file name extensions", which is not the case any more everywhere else? This is probably not an issue with this app, but with the nextcloud core.

My plan for a quick fix is to check before creation if the chosen filename ends with the expected extension, and adding it, should it be missing.

> Renaming a file named without file name extension does not result in it being recognized apparently, the file icon does not change, not even when reloading the page. One actually has to delete and recreate the file.

This I'm not sure about. Maybe I can pass a mime-type when creating the file with the Nextcloud provided dav client, but I have yet to see an example of that.

> Files are created in the wrong folder: one level up! So not inside the folder where one chooses the menu entry to create them. Why?!? One has to move each one afterwards.

You are right about this bug, I was able to reproduce it. Files are simply always created in the root directory. It should only affect Nextcloud 28, but still I will have to release yet another version later today. I will also add a Playwright test, so this doesn't happen again.

> There is no validation of entered links at all. Only when using the files one gets a hint that the file appears to be broken.

This seems like a new feature to me. Links have never been "validated". 

I'm interested to hear how would you validate a link? `foo` can be a link if you run a http service on your `foo` device in your network. As far as I'm concerned people can put into their link files whatever they want 😄 

The one giving you the hint that the link is broken should be your browser. If you're getting an error from the app because of a broken link you put into your link file, this is another bug and unrelated to the link being broken.

Please open a new issue if you can provide more details on this 😊 

----

I will update here when the directory bug is fixed and there's another release out.

**Comment by arkascha** [comment]
> You are right about this bug, I was able to reproduce it. Files are simply always created in the root directory. It should only affect Nextcloud 28, but still I will have to release yet another version later today. I will also add a Playwright test, so this doesn't happen again.

In my case files were not created in the "root" folder (I assume you refer to the user's "root" folder). But simply in the folder one level further up. 


>> There is no validation of entered links at all. Only when using the files one gets a hint that the file appears to be broken.

> This seems like a new feature to me. Links have never been "validated".
That might be the case, I have no idea. 

> I'm interested to hear how would you validate a link? foo can be a link if you run a http service on your foo device in your network. As far as I'm concerned people can put into their link files whatever they want 😄
Well, I'd go and parse out the protocol scheme from whatever has been entered. If recognized, you can make a lookup. In case of a http(s) link for example you can simply make a head request and check the response. If the protocol scheme is _not_ recognized you can point that out to the user and still leave the choice and feature to him.

**Comment by te-online** [comment]
> In my case files were not created in the "root" folder (I assume you refer to the user's "root" folder). But simply in the folder one level further up.

Got it 🙂 Does it work now?

> Well, I'd go and parse out the protocol scheme from whatever has been entered. If recognized, you can make a lookup. In case of a http(s) link for example you can simply make a head request and check the response. If the protocol scheme is not recognized you can point that out to the user and still leave the choice and feature to him.

Thanks for making it more clear. I think that's an interesting feature request. If you'd like we can track this in a new issue.

There may be some challenges to overcome, like what is an exhaustive or even good list of protocols to check for? I like the warning aspect, though. It gives the user a heads-up, but also provides guidance. Requesting the URL could be a potential privacy issue and could cause issues with links that can only be used once, but it might just be something the administrator of the instance or each user can configure...

That being said, I have no interest in expanding the scope of this app at this time. Sorry! You are more than welcome to contribute by forking and opening a PR (preferably early on, so we can discuss these challenges I mentioned) or fork the app and make your own extended version 😊

**Comment by arkascha** [comment]
> > files were not created in the "root" folder
> Got it 🙂 Does it work now?

Works for me now. Thanks for fixing that severe bug! 

> > parse out the protocol scheme
> Thanks for making it more clear. I think that's an interesting feature request. If you'd like we can track this in a new issue.

Is that really a new feature request? 
I see a "Visit Link" button on the form to create a file. That is more or less the same thing. Not sure, has it always been there? Or did you add it now? If it has been there then this is _not_ a new feature request. If you added it then you already introduced the feature. So there is no need for a separate issue / feature request. 

Using a "validation" button is a well known approach to test validity of entered references. It also prevents the points you mentioned where you hesitate, although I consider things like one-time reference extreme edge cases which I personally would _not_ value strong enough to actually block an otherwise sense making feature. Also calling such references cannot somehow violate privacy interests: it is a private form into which the reference is voluntarily entered so that it is saved and used later on. 

I would suggest to enhance that button. "Validate" instead of "Visit link". As a result of such validation you could mark the input field as "validated" (green / thumbs up), that is already the case in nextcloud for example when you configure a TURN-Server. 

As for the protocols: just pick the main few: https, http, ftp, sftp, ftps, mailto. You rely on the browser to support those anyway, so you can simply use the list of protocols supported by mainstream browsers. No voodoo science in that. Yes, that does mean not simply handing over the reference to a browser, but to actually follow it yourself. A simple switch statement for the evaluation. 

I do understand your hesitation. But in my experience this boils down to judging between edge cases on the one hand side and frustrated or annoyed users on the other hand side. Your choice. 

In general I do engage in making contributions. I like that way to collaborate. However I have left behind PHP so many years ago. And it is not really a platform I miss in my daily work, sorry ;-)

**Comment by arkascha** [comment]
It looks like this app is working again as expected! 
Thank you so much for fixing this  issue! 
Issue closed.
